### PR TITLE
feat(parser): support parser keyword this

### PR DESCRIPTION
**Pull request by esatterwhite**
the parser will handle `this` in a lookup value if it is the first
value `#this.foo`. This will change the look up behavior to read data
from the input value passed to the `lookup` function

see: #75

### CODE REVIEW

File: lib/parser/parser.js [hunk 1]
```diff
@@ -40,6 +41,19 @@ class LookupParser extends CstParser {
...
+      // `this` must be first token in a lookup
+      // it cannot appear deep in an object reference
+      this.OPTION({
+        GATE: () => {
+          const token = this.LA(1).tokenType
+          return token === THIS
+        }
+      , DEF: () => {
+          this.CONSUME(THIS)
+          this.CONSUME(DOT) // just #self is not valid
```
Comment (darinspivey) [inline lib/parser/parser.js:53]: This, not self

File: lib/parser/tokens.js [hunk 2]
```diff
@@ -27,6 +27,11 @@ const HASH = createToken({
 , pattern: /#/
 })
 
+const THIS = createToken({
+  name: 'this'
+, pattern: /this/
```
Comment (darinspivey) [inline lib/parser/tokens.js:32]: Shouldn't it have a word boundaries? Wouldn't something like thisthing match?

**Review by darinspivey — COMMENTED** [review]
Just a few initial comments until a deeper review later

File: lib/parser/tokens.js (hunk 2 above)
Comment (esatterwhite) [reply lib/parser/tokens.js:32]: No this is how it wants you to define token patterns generally. adding boundaries just treats it as an `identifier`

File: README.md [hunk 3]
```diff
@@ -348,6 +348,12 @@ chain.lookup('#three.one') // 1
 chain.lookup('#four.2') // 'c'
 chain.lookup('#four.3.d') // 'bleck'
 chain.lookup('!template:"{{#one}}{{#two}}{{#three.one}}"{{#four.0}}"') // '121a'
+chain.lookup({
+  one: 1
+, two: '#self.one'
```
Comment (esatterwhite) [inline README.md:353]: this

File: lib/parser/tokens.js (hunk 2 above)
Comment (esatterwhite) [reply lib/parser/tokens.js:32]: image

File: lib/chain.js [hunk 4]
```diff
@@ -52,6 +53,10 @@ module.exports = class SetupChain {
     }
   }
 
+  get ctx() {
```
Comment (mdeltito) [inline lib/chain.js:56]: nit: `current_context` or `active_context` or something like that?

File: lib/chain.js [hunk 5]
```diff
@@ -80,14 +85,18 @@ module.exports = class SetupChain {
...
     for (const [key, value] of Object.entries(path)) {
 
       // remap array values with lookup result
       if (Array.isArray(value)) {
         out[key] = value.map((val) => {
           if (typeOf(val) === 'object') return this.lookup(val)
-          return this[lookup](this.parse(val))
+          this.context.push(val)
+          const resolved = this[lookup](this.parse(val))
+          this.context.pop()
```
Comment (mdeltito) [inline lib/chain.js:98]: is this even necessary for non-object types? I can't see a path where the context pushed/popped here is actually used.

**Review by mdeltito — COMMENTED** [review]
just a question, I like the syntax and this change looks really clean

File: lib/chain.js (hunk 5 above)
Comment (esatterwhite) [reply lib/chain.js:98]: Maybe not, but it keeps the order of the stack. other wise I think I would have to keep track of the case where item were not added. Like this its just dumb recursion

File: lib/chain.js (hunk 4 above)
Comment (esatterwhite) [reply lib/chain.js]: I flipped them, so `context` is the public-ish facing property.

File: lib/chain.js (hunk 5 above)
Comment (esatterwhite) [reply lib/chain.js]: Oh i see what you mean. don't handle the array specifically, and let the object bits do everything?

Comment (mdeltito) [reply lib/chain.js]: pretty much what you said - I don't see a code path where an array would actually use the context placed on the stack here

**Review by mdeltito — APPROVED** [review]
LGTM

**Review by darinspivey — APPROVED** [review]

**Pull request merged**